jgrapht-core/src/main/java/org/jgrapht/alg/tour/TwoApproxMetricTSP.java

```diff
@@ -0,0 +1,139 @@
+ * satisfy the triangle inequality.
+ * 
+ * <p>
+ * This is an implementation of the folklore algorithm which returns a depth-first ordering of the
+ * minimum spanning tree. The algorithm is a 2-approximation assuming that the instance satisfies
+ * the triangle inequality. The running time is O(|V|^2 log|V|).
+ * 
+ * <p>
+ * See <a href="https://en.wikipedia.org/wiki/Travelling_salesman_problem">wikipedia</a> for more
+ * details.
+ * 
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author Dimitrios Michail
+ */
+public class TwoApproxMetricTSP<V, E>
+    implements TSPAlgorithm<V, E>
+{
+    /**
+     * Construct a new instance
+     */
+    public TwoApproxMetricTSP()
+    {
+    }
+
+    /**
+     * {@inheritDoc}
+     * 
+     * Computes a 2-approximate tour.
+     */
+    @Override
+    public GraphPath<V, E> getTour(UndirectedGraph<V, E> graph)
+    {
+        if (!GraphTests.isComplete(graph)) {
+            throw new IllegalArgumentException("Graph is not complete");
+        }
+        if (graph.vertexSet().isEmpty()) {
+            throw new IllegalArgumentException("Graph contains no vertices");
+        }
+
+        /*
+         * Special case singleton vertex
+         */
+        if (graph.vertexSet().size() == 1) {
+            V start = graph.vertexSet().iterator().next();
+            return new GraphWalk<>(
+                graph, start, start, Collections.singletonList(start), Collections.emptyList(), 0d);
+        }
+
+        /*
+         * Create MST
+         */
+        UndirectedGraph<V, DefaultEdge> mst = new SimpleGraph<>(DefaultEdge.class);
+        for (V v : graph.vertexSet()) {
+            mst.addVertex(v);
+        }
+        for (E e : new KruskalMinimumSpanningTree<>(graph).getSpanningTree().getEdges()) {
+            mst.addEdge(graph.getEdgeSource(e), graph.getEdgeTarget(e));
+        }
+
+        /*
+         * Perform a depth-first-search traversal
+         */
+        int n = graph.vertexSet().size();
+        Set<V> found = new HashSet<>(n);
+        List<V> tour = new ArrayList<>(n);
+        V start = graph.vertexSet().iterator().next();
+        DepthFirstIterator<V, DefaultEdge> dfsIt = new DepthFirstIterator<>(mst, start);
+        while (dfsIt.hasNext()) {
+            V v = dfsIt.next();
+            if (found.add(v)) {
+                tour.add(v);
+            }
+        }
+        // repeat the start vertex
+        tour.add(start);
+
+        /*
+         * Explicitly build the path.
+         */
+        List<E> tourEdges = new ArrayList<E>(n);
+        double tourWeight = 0d;
+        Iterator<V> tourIt = tour.iterator();
+        V u = tourIt.next();
+        while (tourIt.hasNext()) {
+            V v = tourIt.next();
+            E e = graph.getEdge(u, v);
+            tourEdges.add(e);
+            tourWeight += graph.getEdgeWeight(e);
+            u = v;
+        }
+        return new GraphWalk<>(graph, start, start, null, tourEdges, tourWeight);
```

Comment:
Consider returning
`new GraphWalk<>(graph, start, start, tour, tourEdges, tourWeight)`
instead of
`new GraphWalk<>(graph, start, start, null, tourEdges, tourWeight)`

That way it is cheaper to query for instance the city visited at position i in the sequence.